ansible/cloud_providers/gcp_infrastructure_deployment.yml

```diff
@@ -0,0 +1,193 @@
+  connection: local
+  gather_facts: false
+  become: false
+  tags:
+    - step001
+    - deploy_infrastructure
+  tasks:
+    - name: Ensure gcloud is installed
+      environment:
+        PATH: /usr/bin
+      command: which gcloud
+      register: gcloud_result
+
+    - name: Fail if gcloud not available
+      fail:
+        msg: you need Google Cloud SDK installed
+      when: gcloud_result is failed
+
+    - name: Get SSH public key
+      set_fact:
+        ssh_key: "~/.ssh/{{key_name}}.pem"
+        ssh_key_data: "{{lookup('file', '~/.ssh/{{key_name}}.pub')}}"
+
+    - name: Set the destination for the template
+      set_fact:
+        t_dest: "{{output_dir}}/{{ env_type }}.{{ guid }}.{{cloud_provider}}_cloud_template"
+      tags:
+        - gcp_infrastructure_deployment
+        - validate_gcp_template
+        - gen_gcp_template
+
+    - name: Generate GCP deployment manager template
+      template:
+        src: "../configs/{{ env_type }}/files/cloud_providers/{{cloud_provider}}_cloud_template.j2"
+        dest: "{{t_dest}}"
+      tags:
+        - gcp_infrastructure_deployment
+        - validate_gcp_template
+        - gen_gcp_template
+
+    # use command line 'gcloud' to validate template and deploy
+    - name: Login to Google Cloud SDK
+      command: >-
+        gcloud auth
+        activate-service-account "{{ gcp_account }}"
+        --key-file="{{ gcp_credentials_file }}"
+        --project="{{ gcp_project_id }}"
+      environment:
+        PATH: /usr/bin
+      tags:
+        - validate_gcp_template
+        - create_inventory
+        - must
+
+    - name: Get SSH public key
+      set_fact:
+        ssh_key: "~/.ssh/{{key_name}}.pem"
+        ssh_key_data: "{{lookup('file', '~/.ssh/{{key_name}}.pub')}}"
+      tags:
+        - validate_gcp_template
+        - must
+        - create_inventory
+
+    - name: Starting Google deployment-manager with template
+      environment:
+        PATH: /usr/bin
+      command: >-
+        gcloud deployment-manager deployments
+        create "{{ project_tag }}-base"
+        --config "{{ t_dest }}"
+      register: gcp_deploy
+      tags:
+        - gcp_infrastructure_deployment
+        - gcp_deployment_manager
+      until: gcp_deploy is succeeded
+      retries: 0
+
+    - debug:
+        var: gcp_deploy
+        verbosity: 2
+      tags:
+        - gcp_infrastructure_deployment
+
+    - name: Run infra-gcp-create-inventory Role
+      import_role:
+        name: infra-gcp-create-inventory
+
+    - name: Fetch DNS zone Info
+      gcp_dns_managed_zone_facts:
+        dns_name: '{{ cluster_dns_zone + "."}}'
+        project: "{{ gcp_project_id }}"
+        auth_kind: "{{ gcp_auth_type }}"
+        service_account_file: "{{ gcp_credentials_file }}"
+      register: gcp_managed_zone
+      tags:
+        - gcp_infrastructure_deployment
+      when:
+        - HostedZoneId != "none"
+
+    - name: Add delegation for NS to the main DNSZone
+      gcp_dns_resource_record_set:
```

Comment:
This can't be part of the cloud template ?

Reply:
I modeled after the Azure cloud provider so I didn't consider that, I believe I can move this into the template.

Reply:
There is no clean way to include DNS records in the deployment manager template. I can include a third party script but then we would also need to branch that script and track changes to it. At least the way it is now it uses Ansible so its stateless.

Reply:
After more research, Google doesn't update DNS in templates anywhere its its sample library without using a custom python script which is something else we'd need to watch and version.

Reply:
OK, it's OK for now.